{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import nbformat\n",
    "import sys\n",
    "import textwrap\n",
    "import xlrd\n",
    "filename='sample_data.xlsx'\n",
    "data=xlrd.open_workbook(filename)\n",
    "data_sheet=data.sheet_by_index(0)\n",
    "\n",
    "\n",
    "\n",
    "for i in range(1,data_sheet.nrows):\n",
    "    nb = nbformat.v4.new_notebook()\n",
    "    \n",
    "    \n",
    "    gse=data_sheet.cell(i,0).value\n",
    "    gse=gse.encode('ascii','ignore')\n",
    "    \n",
    "    platform=data_sheet.cell(i,1).value\n",
    "    platform=platform.encode('ascii','ignore')\n",
    "    \n",
    "    sample_id=gse+\"_\"+platform\n",
    "    \n",
    "    control_samples_excel=data_sheet.cell(i,2).value.encode('ascii','ignore')\n",
    "    experimental_samples_excel=data_sheet.cell(i,3).value.encode('ascii','ignore')\n",
    "    \n",
    "    number_of_samples=len(control_samples_excel.split(\"|\"))\n",
    "      \n",
    "\n",
    "    source_0 = textwrap.dedent(\"\"\"\\\n",
    "        from IPython.display import HTML\n",
    "        HTML('''<script>\n",
    "        code_show=true; \n",
    "        function code_toggle() {\n",
    "         if (code_show){\n",
    "         $('div.input').hide();\n",
    "         } else {\n",
    "         $('div.input').show();\n",
    "         }\n",
    "         code_show = !code_show\n",
    "        } \n",
    "        $( document ).ready(code_toggle);\n",
    "        </script>\n",
    "        <form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')       \n",
    "        \"\"\")\n",
    "    code_cell_0 = nbformat.v4.new_code_cell(source=source_0)\n",
    "    nb.cells.append(code_cell_0)\n",
    "    \n",
    "   \n",
    "    \n",
    "        \n",
    "    markdown_cell_1_content = \"## Exploratory analysis of dataset\" + \" \" + gse + \" \" + \"obtained from Gene Expression Omnibus (GEO)\"\n",
    "    markdown_cell_1 = nbformat.v4.new_markdown_cell(markdown_cell_1_content)\n",
    "    nb.cells.append(markdown_cell_1)\n",
    "\n",
    "    markdown_cell_2_content = \"The data discussed in this publication have been deposited in NCBI's Gene Expression Omnibus and are accessible through GEO Series accession number\" + \" \" + gse+ \" \" + \"and are available [here](at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=).\"\n",
    "\n",
    "    markdown_cell_2 = nbformat.v4.new_markdown_cell(markdown_cell_2_content)\n",
    "    nb.cells.append(markdown_cell_2)\n",
    "\n",
    "    markdown_cell_3_content = textwrap.dedent(\"\"\"\\\n",
    "        ### Goals of this report\n",
    "        This report aims to conduct an exploratory analysis of the RNA-seq data generated by this study using an [RNA-seq pipeline](https://f1000researchdata.s3.amazonaws.com/manuscripts/9804/8fa2f4e7-d0f3-488c-85f1-0b3a9816faef_9110_-_avi_maaayn.pdf?doi=10.12688/f1000research.9110.1) developed by [the Ma'ayan Lab](http://labs.icahn.mssm.edu/maayanlab/) at [the Icahn School of Medicine at Mount Sinai](http://icahn.mssm.edu/). This [Jupyter Notebook](http://jupyter.org/) was automatically generated <a href='#Citations'>[1]</a>.\n",
    "        \"\"\")\n",
    "    markdown_cell_3 = nbformat.v4.new_markdown_cell(markdown_cell_3_content)\n",
    "    nb.cells.append(markdown_cell_3)\n",
    "\n",
    "    markdown_cell_4_content = textwrap.dedent(\"\"\"\\\n",
    "        ### Analysis Overview\n",
    "        #### 1.  <a href='#Dimensionality Reduction'>Dimensionality Reduction</a> \n",
    "        Dimensionality reduction helps analyze and visualize multidimensional data by reducing the data to 3D space. Principal component analysis (PCA) was applied to the data in this analysis.\n",
    "        #### 2.  <a href='#Interactive Heatmap'>Interactive Heatmap</a>\n",
    "        [Clustergrammer](http://clustergrammer.readthedocs.io/clustergrammer_widget.html) is an interactive tool that can be used to visualize a wide range of biological data. It is particularly useful for gene expression data. Clusters of upregulated and downregulated genes for the groups are clearly visible, which simplifies further analysis and comparisons between differential expression signatures.\n",
    "        #### 3.  <a href='#Differential Expression Analysis'>Differential Expression Analysis</a>\n",
    "        This analysis helps to identify genes that show differences in expression as compared to control. This information can be used to obtain a deeper understanding of the molecular mechanisms of biological processes.\n",
    "        #### 4.  <a href='#Gene Set Enrichment Analysis'>Gene Set Enrichment Analysis</a>\n",
    "        Gene Set Enrichment Analysis uses the upregulated and downregulated genes identified in differential expression analysis to provide valuable insight into the underlying mechanisms of biological processes.\n",
    "        #### 5.  <a href='#Small Molecule Predictions'>Small Molecule Predictions</a>\n",
    "        This analysis produces a set of small molecules that are predicted either mimic or reverse the expression profiles observed in the experiment.\n",
    "\n",
    "        \"\"\")\n",
    "    markdown_cell_4 = nbformat.v4.new_markdown_cell(markdown_cell_4_content)\n",
    "    nb.cells.append(markdown_cell_4)\n",
    "\n",
    "    markdown_cell_5 = nbformat.v4.new_markdown_cell(\"<a id='Dimensionality Reduction'></a>\")\n",
    "    nb.cells.append(markdown_cell_5)\n",
    "\n",
    "    markdown_cell_6_content = textwrap.dedent(\"\"\"\\\n",
    "        ## 1. Dimensionality Reduction\n",
    "\n",
    "        ### Principal Component Analysis (PCA)\n",
    "        PCA was used to reduce the data to three-dimensional space.Counts per million (CPM) and z-score normalization methods were applied to the data. A PCA plot is expected to show clusters of samples relevant to experimental groups <a href='#Citations'>[2]</a>.\n",
    "        \"\"\")\n",
    "    markdown_cell_6 = nbformat.v4.new_markdown_cell(markdown_cell_6_content)\n",
    "    nb.cells.append(markdown_cell_6)\n",
    "\n",
    "    import textwrap\n",
    "\n",
    "    \n",
    "    source_variables=textwrap.dedent(\"\"\"\\\n",
    "\n",
    "    #%store -r gse\n",
    "    #%store -r platform\n",
    "    #%store -r control_samples_excel\n",
    "    #%store -r experimental_samples_excel\n",
    "    #%store -r number_of_samples\n",
    "    #%store -r sample_id\n",
    "\n",
    "    gse=\"{gse}\"\n",
    "    platform=\"{platform}\"\n",
    "    sample_id=\"{sample_id}\"\n",
    "    control_samples_excel=\"{control_samples_excel}\"\n",
    "    experimental_samples_excel=\"{experimental_samples_excel}\"\n",
    "\n",
    "\n",
    "    \"\"\".format(**locals()))\n",
    "\n",
    "\n",
    "    code_cell_variables=nbformat.v4.new_code_cell(source=source_variables)\n",
    "    nb.cells.append(code_cell_variables)\n",
    "\n",
    "\n",
    "    source_number_of_samples=\"number_of_samples=({})\".format(number_of_samples)\n",
    "\n",
    "    code_cell_number_of_samples=nbformat.v4.new_code_cell(source=source_number_of_samples)\n",
    "    nb.cells.append(code_cell_number_of_samples)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    source_1 = textwrap.dedent(\"\"\"\\\n",
    "        %%capture\n",
    "\n",
    "        # Initialize Plotly and R magic\n",
    "        %load_ext rpy2.ipython\n",
    "\n",
    "        %R require(limma)\n",
    "        %R require(edgeR)\n",
    "        %R require(ggpot)\n",
    "        %R require(gridExtra)\n",
    "        from plotly.offline import init_notebook_mode\n",
    "        init_notebook_mode()\n",
    "\n",
    "        import sys\n",
    "        import xlrd\n",
    "        sys.path.append('scripts')\n",
    "        sys.path.append('gloriascripts')\n",
    "        import pandas as pd\n",
    "        import archs4\n",
    "        import numpy as np \n",
    "        \n",
    "        \n",
    "    \n",
    "        \n",
    "        \n",
    "        #platform=platform_nb\n",
    "        #control_samples_excel=control_samples_excel_nb\n",
    "        #experimental_samples_excel=experimental_samples_excel_nb\n",
    "        #number_of_samples=number_of_samples_nb\n",
    "        #sample_id=sample_id_nb\n",
    "        \n",
    "        #sample_id\n",
    "        #platform\n",
    "        #control_samples_excel\n",
    "        #experimental_samples_excel\n",
    "        #number_of_samples\n",
    "        \n",
    "        \"\"\")\n",
    "\n",
    "    code_cell_1 = nbformat.v4.new_code_cell(source=source_1)\n",
    "    nb.cells.append(code_cell_1)\n",
    "\n",
    "    source_10=\"import xlrd\"\n",
    "    code_cell_10 = nbformat.v4.new_code_cell(source=source_10)\n",
    "    nb.cells.append(code_cell_10)\n",
    "\n",
    "\n",
    "#     source_4 = \"rawcount_dataframe,sample_title_dict,treatment_dataframe = archs4.fetch_dataset('{}')\".format(sample_id)\n",
    "\n",
    "#     code_cell_4 = nbformat.v4.new_code_cell(source=source_4)\n",
    "#     nb.cells.append(code_cell_4)\n",
    "\n",
    "    source_4 = \"rawcount_dataframe,sample_title_dict,treatment_dataframe = archs4.fetch_dataset(sample_id)\"\n",
    "\n",
    "    code_cell_4 = nbformat.v4.new_code_cell(source=source_4)\n",
    "    nb.cells.append(code_cell_4)\n",
    "\n",
    "\n",
    "\n",
    "    source_5 = textwrap.dedent(\"\"\"\\\n",
    "    \n",
    "        rawcount_dataframe=rawcount_dataframe.drop(rawcount_dataframe.index[rawcount_dataframe.mean(axis=1)<=rawcount_dataframe.shape[1]])\n",
    "        rawcount_dataframe_original=rawcount_dataframe.copy(deep=True)\n",
    "\n",
    "        group_dataframe=pd.DataFrame()\n",
    "\n",
    "        group_dataframe_list=sample_title_dict.values()\n",
    "\n",
    "        group_dataframe['treatment']=pd.Series(sample_title_dict.get(v) for v in rawcount_dataframe_original.columns)\n",
    "        group_dataframe['treatment']=pd.Series(v.split(' (GSM')[0] for v in group_dataframe['treatment'])\n",
    "        treatment_dataframe.columns=['treatment']\n",
    "\n",
    "        group_dataframe.index=treatment_dataframe['treatment']\n",
    "        treatment_dataframe.index=group_dataframe['treatment']\n",
    "\n",
    "        rawcount_dataframe.columns=group_dataframe['treatment']\n",
    "\n",
    "\n",
    "        norm_dataframe=pd.DataFrame()\n",
    "        norm_dataframe=rawcount_dataframe.copy(deep=True)\n",
    "\n",
    "\n",
    "\n",
    "        #cpm normalizing the columns\n",
    "\n",
    "        cpm_factors=rawcount_dataframe.sum(axis=0)/1000000\n",
    "        for i in range(0,rawcount_dataframe.shape[1]):\n",
    "            norm_dataframe.iloc[:,i]=rawcount_dataframe.iloc[:,i]/cpm_factors[i]\n",
    "\n",
    "        #z score normalize the rows\n",
    "\n",
    "        from scipy import stats\n",
    "        for rows in range(0,rawcount_dataframe.shape[0]):\n",
    "            norm_dataframe.iloc[rows,:] = stats.zscore(rawcount_dataframe.iloc[rows,:])\n",
    "\n",
    "        from sklearn.decomposition import PCA\n",
    "        pca_3=PCA(n_components=3)\n",
    "        pca_3.fit_transform(norm_dataframe)\n",
    "\n",
    "        import plot_3d_scatter\n",
    "        fig=plot_3d_scatter.plot_3d_scatter(pca_3.components_[0],pca_3.components_[1],pca_3.components_[2], size=5, color_by=treatment_dataframe['treatment'].tolist(),\n",
    "        xlab= 'PC1 ({:.2%} of variance explained)'.format(pca_3.explained_variance_ratio_[0]),\n",
    "        ylab= 'PC2 ({:.2%} of variance explained)'.format(pca_3.explained_variance_ratio_[1]),\n",
    "        zlab= 'PC3 ({:.2%}of variance explained)'.format(pca_3.explained_variance_ratio_[2]))\n",
    "\n",
    "\n",
    "        \"\"\")\n",
    "\n",
    "    code_cell_5 = nbformat.v4.new_code_cell(source=source_5)\n",
    "    nb.cells.append(code_cell_5)\n",
    "\n",
    "\n",
    "   # Plotting PCA\n",
    "\n",
    "    source_2 = textwrap.dedent(\"\"\"\\\n",
    "            import plotly.offline as py\n",
    "            from plotly.offline import iplot,init_notebook_mode\n",
    "            init_notebook_mode()\n",
    "            py.iplot(fig)\n",
    "        \"\"\")\n",
    "\n",
    "    code_cell_2 = nbformat.v4.new_code_cell(source=source_2)\n",
    "    nb.cells.append(code_cell_2)\n",
    "\n",
    "    source_5 = textwrap.dedent(\"\"\"\\\n",
    "\n",
    "        %Rpush rawcount_dataframe\n",
    "        %Rpush treatment_dataframe\n",
    "        %Rpush group_dataframe\n",
    "        %R design<-model.matrix(~1+treatment, group_dataframe)\n",
    "        %R voom_dataframe<-voom(rawcount_dataframe,plot=FALSE)\n",
    "        %R voom_dataframe<-as.data.frame(voom_dataframe)\n",
    "        %Rpull voom_dataframe\n",
    "        \"\"\")\n",
    "\n",
    "    code_cell_5 = nbformat.v4.new_code_cell(source=source_5)\n",
    "    nb.cells.append(code_cell_5)\n",
    "\n",
    "\n",
    "    markdown_cell_9_content = textwrap.dedent(\"\"\"\\\n",
    "        #### Figure 1 | PCA analysis of the RNA-seq dataset. \n",
    "        The plot depicts the PCA analysis of the dataset. Each point on the graph represents a single sample. The axis represent the principal components, which are linear combinations of the variables from each expression vector. The variance accounted for by each principal component can be seen on the respective axis.\n",
    "        \"\"\")\n",
    "    markdown_cell_9 = nbformat.v4.new_markdown_cell(markdown_cell_9_content)\n",
    "    nb.cells.append(markdown_cell_9)\n",
    "\n",
    "\n",
    "    #Clustergrammer\n",
    "\n",
    "    markdown_cell_7 = nbformat.v4.new_markdown_cell(\"<a id='Interactive Heatmap'></a>\")\n",
    "    nb.cells.append(markdown_cell_7)\n",
    "\n",
    "    markdown_cell_8_content = textwrap.dedent(\"\"\"\\\n",
    "    ## 2. Interactive Heatmap\n",
    "\n",
    "    500 most variable genes were visualized in this [Clustergrammer](http://clustergrammer.readthedocs.io/clustergrammer_widget.html) <a href='#Citations'>[3]</a>.The names of the samples can be seen as column labels; you can zoom in to see the gene names as rows. The tool is interactive; you can arrange row and column orders by using the panel to the left of the heatmap. The cluster of interest can be selected and the resulting genes can be sent to [Enrichr](http://amp.pharm.mssm.edu/Enrichr/) for further analysis. The clustergrammer instance can be shared and the matrix can be downloaded as a .txt file.\n",
    "   \n",
    "    \"\"\")\n",
    "    markdown_cell_8 = nbformat.v4.new_markdown_cell(markdown_cell_8_content)\n",
    "    nb.cells.append(markdown_cell_8)\n",
    "\n",
    "    source_3 = textwrap.dedent(\"\"\"\\\n",
    "        import plot_clustergrammer\n",
    "        plot_clustergrammer.display(voom_dataframe) \n",
    "        \"\"\")\n",
    "\n",
    "    code_cell_3 = nbformat.v4.new_code_cell(source=source_3)\n",
    "    nb.cells.append(code_cell_3)\n",
    "\n",
    "    markdown_cell_5_content = textwrap.dedent(\"\"\"\\\n",
    "        #### Figure 2 | Interactive heatmap for top 500 most variable genes.\n",
    "        The clustergrammer depicts the normalized expression levels of top 500 most variable genes. Voom method in R was used to normalize the data. The columns represent the sample names. You can zoom in to see the gene names as rows.\n",
    "        \"\"\")\n",
    "    markdown_cell_5 = nbformat.v4.new_markdown_cell(markdown_cell_5_content)\n",
    "    nb.cells.append(markdown_cell_5)\n",
    "\n",
    "    markdown_cell_12 = nbformat.v4.new_markdown_cell(\"<a id='Gene Co-Expression Analysis'></a>\")\n",
    "    nb.cells.append(markdown_cell_12)\n",
    "\n",
    "\n",
    "    #EXAMPLE\n",
    "    #source_9 = \"data=xlrd.open_workbook('{}')\".format(filename)\n",
    "\n",
    "\n",
    "    source_6 = textwrap.dedent(\"\"\"\\\n",
    "        from rpy2.robjects import r, pandas2ri\n",
    "        import pandas as pd\n",
    "        import sys\n",
    "\n",
    "        r.source('scripts/signature.R')\n",
    "        pandas2ri.activate()\n",
    "    \"\"\")\n",
    "    code_cell_6 = nbformat.v4.new_code_cell(source=source_6)\n",
    "    nb.cells.append(code_cell_6)\n",
    "\n",
    "    source_ctrl=\"control_samples=list(control_samples_excel.split('|')[v] for v in range(number_of_samples))\"\n",
    "\n",
    "    code_cell_ctrl=nbformat.v4.new_code_cell(source=source_ctrl)\n",
    "    nb.cells.append(code_cell_ctrl)\n",
    "\n",
    "\n",
    "    source_exp=\"experimental_samples=list(experimental_samples_excel.split('|')[v] for v in range(number_of_samples) )\"\n",
    "\n",
    "    code_cell_exp=nbformat.v4.new_code_cell(source=source_exp)\n",
    "    nb.cells.append(code_cell_exp)\n",
    "\n",
    "\n",
    "\n",
    "    source_20= textwrap.dedent(\"\"\"\\\n",
    "        import gene_signature\n",
    "        limma_dataframe=gene_signature.signature(rawcount_dataframe_original,method = 'limma', experimental_samples=experimental_samples, control_samples=control_samples)\n",
    "        limma_dataframe['significant']=[rowData['adj.P.Val'] < 0.2 and abs(rowData['logFC']) > 1 for index, rowData in limma_dataframe.iterrows()]\n",
    "\n",
    "        sign=limma_dataframe[limma_dataframe['significant']==True]\n",
    "        insign=limma_dataframe[limma_dataframe['significant']==False]\n",
    "\n",
    "        import numpy as np\n",
    "        import customplot\n",
    "        \"\"\")\n",
    "\n",
    "    code_cell_20 = nbformat.v4.new_code_cell(source=source_20)\n",
    "    nb.cells.append(code_cell_20)\n",
    "\n",
    "\n",
    "    #Differential Expression Analysis\n",
    "\n",
    "    markdown_cell_18 = nbformat.v4.new_markdown_cell(\"<a id='Differential Expression Analysis'></a>\")\n",
    "    nb.cells.append(markdown_cell_18)\n",
    "\n",
    "    markdown_cell_17_content = textwrap.dedent(\"\"\"\\\n",
    "         ## 3. Differential Expression Analysis\n",
    "        \"\"\")\n",
    "    markdown_cell_17 = nbformat.v4.new_markdown_cell(markdown_cell_17_content)\n",
    "    nb.cells.append(markdown_cell_17)\n",
    "\n",
    "    markdown_cell_13_content = textwrap.dedent(\"\"\"\\\n",
    "        ### 3.1 Volcano and MA Plots\n",
    "\n",
    "        [Limma package](https://bioconductor.org/packages/release/bioc/html/limma.html) in R was used to obtain differential gene expression signatures <a href='#Citations'>[4]</a>. \n",
    "\n",
    "        A volcano plot visualizes the results of gene expression analysis and depicts the logarithm of P-values (-log10(P-value)) on the y-axis versus log-transformed fold changes (logFC) on the x-axis.\n",
    "\n",
    "        MA plot depicts the log-transformed fold changes (M) for each gene versus its average expression (A). Fold changes, P-values and average expressions were calculated using limma package in R.\n",
    "\n",
    "        \"\"\")\n",
    "\n",
    "    markdown_cell_13 = nbformat.v4.new_markdown_cell(markdown_cell_13_content)\n",
    "    nb.cells.append(markdown_cell_13)\n",
    "\n",
    "    source_18=textwrap.dedent(\"\"\"\\\n",
    "\n",
    "    volcano_df=df_sign_up = pd.DataFrame({'logFC': limma_dataframe['logFC'],'adj.P.Val':limma_dataframe['adj.P.Val'],'p.Value':limma_dataframe['P.Value'],'gene':limma_dataframe.index,'AveExpr' : limma_dataframe['AveExpr']})\n",
    "\n",
    "    \"\"\")\n",
    "\n",
    "    code_cell_18 = nbformat.v4.new_code_cell(source=source_18)\n",
    "    nb.cells.append(code_cell_18)\n",
    "\n",
    "\n",
    "    source_7 = textwrap.dedent(\"\"\"\\\n",
    "       %%R -i volcano_df -w 900 -h 480 -u px\n",
    "        library(ggplot2)\n",
    "        library(ggrepel)\n",
    "        volcano_df<-as.data.frame(volcano_df)\n",
    "        Genes<-as.factor(abs(volcano_df$logFC) > 1 & volcano_df$adj.P.Val<0.2)\n",
    "        levels(Genes) <- sub(\"^TRUE$\", \"Significant\",levels(Genes))\n",
    "        levels(Genes) <- sub(\"^FALSE$\", \"Not Significant\",levels(Genes))\n",
    "        g_volcano <-ggplot(data=volcano_df, aes(x=logFC, y=-log10(p.Value)),na.rm=TRUE) + geom_point(aes(col=Genes),size=4) + scale_color_manual(values=c(\"black\", \"red\"))+ theme(panel.background = element_blank()) + xlab('Log-tranformed fold changes') + ylab('-log10 (P-Value)') + theme_bw()+theme(axis.text=element_text(size=12),axis.title=element_text(size=14)) + guides(fill=guide_legend(title=NULL)) +ggtitle(\"Volcano Plot\")+theme(plot.title = element_text(family=\"bold\", size=18,hjust=0.5))\n",
    "        set.seed(42)\n",
    "        g_ma<-ggplot(data=volcano_df, aes(x=AveExpr, y=logFC)) + geom_point(aes(col=Genes),size=4) + scale_color_manual(values=c(\"black\", \"red\"))+ theme(panel.background = element_blank())+ xlab('Average gene expression') + ylab('Log-tranformed fold changes') + theme_bw()+theme(axis.text=element_text(size=12),axis.title=element_text(size=14)) + guides(fill=guide_legend(title=NULL)) + ggtitle(\"MA Plot\") +theme(plot.title = element_text(family=\"bold\", size=18,hjust = 0.5))\n",
    "\n",
    "        grid.arrange(g_volcano, g_ma, ncol=2)\n",
    "\n",
    "        \"\"\")\n",
    "    code_cell_7 = nbformat.v4.new_code_cell(source=source_7)\n",
    "    nb.cells.append(code_cell_7)\n",
    "\n",
    "\n",
    "    markdown_cell_14_content = textwrap.dedent(\"\"\"\\\n",
    "        #### Figure 3 | Volcano and MA plots for the RNA-seq dataset.  \n",
    "        Volcano plot depicts the log of P-value for each gene verus log-transformed fold changes as compared to the control samples. MA plot depicts the log-transformed fold changes (M) for each gene versus its average expression (A). Fold changes, average expressions and P-values were calculated using limma package in R. Limma method along with the thresholds (|logFC| > 1 and adjusted P-value < 0.2) were used to identify the significant genes.\n",
    "        \"\"\")\n",
    "    markdown_cell_14 = nbformat.v4.new_markdown_cell(markdown_cell_14_content)\n",
    "    nb.cells.append(markdown_cell_14)\n",
    "\n",
    "\n",
    "\n",
    "    markdown_cell_19_content = textwrap.dedent(\"\"\"\\\n",
    "         ### 3.2 Further exploration of significant genes identified by limma\n",
    "\n",
    "         The following tables show significant upregulated genes and downregulated genes identifies by limma. You can scroll and search for genes of interest by using the toolbar in the tables.\n",
    "        \"\"\")\n",
    "    markdown_cell_19 = nbformat.v4.new_markdown_cell(markdown_cell_19_content)\n",
    "    nb.cells.append(markdown_cell_19)\n",
    "\n",
    "    source_12 = textwrap.dedent(\"\"\"\\\n",
    "        list_sign_up=limma_dataframe[(limma_dataframe['logFC']>0) & (limma_dataframe['significant']==True)].index\n",
    "        list_sign_dn=limma_dataframe[(limma_dataframe['logFC']<0) & (limma_dataframe['significant']==True)].index\n",
    "\n",
    "        df_sign_up = pd.DataFrame({'Significant Upregulated Genes':list_sign_up, 'Log-transformed fold changes': limma_dataframe.loc[list_sign_up]['logFC'],'Adjusted P-value':limma_dataframe.loc[list_sign_up]['adj.P.Val']})\n",
    "        df_sign_up=df_sign_up[['Significant Upregulated Genes','Log-transformed fold changes','Adjusted P-value']]\n",
    "\n",
    "\n",
    "        df_sign_dn = pd.DataFrame({'Significant Downregulated Genes':list_sign_dn, 'Log-transformed fold changes': limma_dataframe.loc[list_sign_dn]['logFC'],'Adjusted P-value':limma_dataframe.loc[list_sign_dn]['adj.P.Val']})\n",
    "        df_sign_dn=df_sign_dn[['Significant Downregulated Genes','Log-transformed fold changes','Adjusted P-value']]\n",
    "\n",
    "        import qgrid\n",
    "        qgrid.nbinstall(overwrite=True)\n",
    "        qgrid.set_defaults(show_toolbar=True, remote_js=True,export_mode=True)       \n",
    "        \"\"\")\n",
    "    code_cell_12 = nbformat.v4.new_code_cell(source=source_12)\n",
    "    nb.cells.append(code_cell_12)\n",
    "\n",
    "\n",
    "\n",
    "    source_13=textwrap.dedent(\"\"\"\\\n",
    "    df_sign_up.index=range(1,len(df_sign_up)+1)\n",
    "    qgrid.show_grid(df_sign_up) \n",
    "    \"\"\")\n",
    "    code_cell_13 = nbformat.v4.new_code_cell(source=source_13)\n",
    "    nb.cells.append(code_cell_13)\n",
    "\n",
    "    markdown_cell_20_content = textwrap.dedent(\"\"\"\\\n",
    "         #### Table 1 | Significant upregulated genes identified by limma as compared to control samples\n",
    "        The table shows the list of significant upregulated genes as compared to control samples. Limma method in R along with the thresholds (logFC > 1 and adjusted P-value < 0.2) were used to identify the significant genes.\n",
    "        \"\"\")\n",
    "    markdown_cell_20 = nbformat.v4.new_markdown_cell(markdown_cell_20_content)\n",
    "    nb.cells.append(markdown_cell_20)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    source_14=textwrap.dedent(\"\"\"\\\n",
    "    df_sign_dn.index=range(1,len(df_sign_dn)+1)\n",
    "    qgrid.show_grid(df_sign_dn) \n",
    "    \"\"\")\n",
    "    code_cell_14 = nbformat.v4.new_code_cell(source=source_14)\n",
    "    nb.cells.append(code_cell_14)\n",
    "\n",
    "    markdown_cell_21_content = textwrap.dedent(\"\"\"\\\n",
    "         #### Table 2 | Significant downregulated genes identified by limma as compared to control samples\n",
    "        The table shows the list of significant downregulated genes as compared to control samples. Limma method in R along with the thresholds (logFC < -1 and adjusted P-value < 0.2) were used to identify the significant genes.\n",
    "        \"\"\")\n",
    "    markdown_cell_21 = nbformat.v4.new_markdown_cell(markdown_cell_21_content)\n",
    "    nb.cells.append(markdown_cell_21)\n",
    "\n",
    "    markdown_cell_26=nbformat.v4.new_markdown_cell(\"<a id='Gene Set Enrichment Analysis'></a>\")\n",
    "    nb.cells.append(markdown_cell_26)\n",
    "\n",
    "\n",
    "    #Gene Set Enrichment Analysis\n",
    "\n",
    "\n",
    "    markdown_cell_22_content = textwrap.dedent(\"\"\"\\\n",
    "        ## 4. Gene Set Enrichment Analysis\n",
    "\n",
    "        The differential gene expression signatures computed with limma are now used for enrichment analysis on top 500 upregulated and downregulated genes using [Enrichr](http://amp.pharm.mssm.edu/Enrichr/). \n",
    "\n",
    "        [Enrichr](http://amp.pharm.mssm.edu/Enrichr/) is a comprehensive gene set enrichment analysis tool. Containing 180,184 annotated gene sets from 102 gene set libraries, it allows for easy and effective search and provides interactive visualizations of enrichment results <a href='#Citations'>[5]</a>. \n",
    "\n",
    "        \"\"\")\n",
    "    markdown_cell_22 = nbformat.v4.new_markdown_cell(markdown_cell_22_content)\n",
    "    nb.cells.append(markdown_cell_22)\n",
    "\n",
    "\n",
    "\n",
    "    source_cd_1=textwrap.dedent(\"\"\"\\\n",
    "    limma_up=limma_dataframe.loc[limma_dataframe['logFC']>=0]\n",
    "    limma_dn=limma_dataframe.loc[limma_dataframe['logFC']<0]\n",
    "    top_500=limma_up['adj.P.Val'].sort_values(ascending=True)[:500].index\n",
    "    bottom_500=limma_dn['adj.P.Val'].sort_values(ascending=True).tail(500).index\n",
    "\n",
    "    #top_500_var_indices=rawcount_dataframe.var(axis=1).sort_values(ascending=False)[:100].index\n",
    "    #top_500_var_genes=rawcount_dataframe.loc[top_500_var_indices]\n",
    "\n",
    "    #top_variable_genes=rawcount_dataframe_original.loc[rawcount_dataframe_original.apply(np.var,1).sort_values(ascending=False)[:5000].index]\n",
    "\n",
    "    #cd_dataframe=gene_signature.signature(top_variable_genes,method = 'CD',experimental_samples = experimental_samples, control_samples = control_samples)\n",
    "\n",
    "    #cd_dataframe.sort_values('CD',axis=0,ascending=False)[:200].index\n",
    "    #top_500=cd_dataframe.sort_values('CD',axis=0,ascending=False)[:200].index\n",
    "    #bottom_500=cd_dataframe.sort_values('CD',axis=0,ascending=False).tail(200).index\n",
    "\n",
    "\n",
    "    \"\"\")\n",
    "    code_cell_cd_1=nbformat.v4.new_code_cell(source=source_cd_1)\n",
    "    nb.cells.append(code_cell_cd_1)\n",
    "\n",
    "    source_cd_2=textwrap.dedent(\"\"\"\\\n",
    "\n",
    "    import upload_genesets\n",
    "    dict_up=upload_genesets.enrichr(top_500)[1]\n",
    "    dict_dn=upload_genesets.enrichr(bottom_500)[1]\n",
    "    user_list_id_up = dict_up.get('userListId')\n",
    "    user_list_id_dn=dict_dn.get('userListId')\n",
    "\n",
    "    import enrichr_library\n",
    "    #up_chea,dn_chea=enrichr_library.data(gene_set_library='ChEA_2016',user_list_id_up=user_list_id_up,user_list_id_dn=user_list_id_dn)\n",
    "    #up_bio_process,dn_bio_process=enrichr_library.data(gene_set_library='GO_Biological_Process_2017b',user_list_id_up=user_list_id_up,user_list_id_dn=user_list_id_dn)\n",
    "    #up_kegg,dn_kegg=enrichr_library.data(gene_set_library='KEGG_2016',user_list_id_up=user_list_id_up,user_list_id_dn=user_list_id_dn)\n",
    "\n",
    "    enrichr_url_up=upload_genesets.enrichr(top_500)[0]\n",
    "    enrichr_url_dn=upload_genesets.enrichr(bottom_500)[0]\n",
    "    \"\"\")\n",
    "    code_cell_cd_2=nbformat.v4.new_code_cell(source=source_cd_2)\n",
    "    nb.cells.append(code_cell_cd_2)\n",
    "\n",
    "    source_17=textwrap.dedent(\"\"\"\\\n",
    "    def make_clickable(val):\n",
    "        return '<a href=\"{}\">{}</a>'.format(val,val)\n",
    "    df = pd.DataFrame([enrichr_url_up,enrichr_url_dn])\n",
    "    df.index=['Upregulated genes','Downregulated genes']\n",
    "    df.columns=['Links to Enrichr results']\n",
    "    df.style.format(make_clickable)\n",
    "    \"\"\")\n",
    "    code_cell_17 = nbformat.v4.new_code_cell(source=source_17)\n",
    "    nb.cells.append(code_cell_17)\n",
    "\n",
    "\n",
    "    markdown_cell_23_content = textwrap.dedent(\"\"\"\\\n",
    "         #### Table 3 | Links to  Enrichr results\n",
    "        The table includes the links to  Enrichr results for top 500 upregulated and downregulated genes identified by limma.\n",
    "        \"\"\")\n",
    "    markdown_cell_23 = nbformat.v4.new_markdown_cell(markdown_cell_23_content)\n",
    "    nb.cells.append(markdown_cell_23)\n",
    "\n",
    "\n",
    "    markdown_cell_29=nbformat.v4.new_markdown_cell(\"<a id='Small Molecule Predictions'></a>\")\n",
    "    nb.cells.append(markdown_cell_29)\n",
    "\n",
    "    #Small molecule predictions\n",
    "\n",
    "\n",
    "    markdown_cell_24_content = textwrap.dedent(\"\"\"\\\n",
    "         ### 5. Small Molecule Predictions\n",
    "\n",
    "        [L1000CDS<sup>2</sup>](http://amp.pharm.mssm.edu/L1000CDS2/#/index) is a state-of-the-art web-based search engine, which provides prioritization of thousands of small-molecule signatures, and their pairwise combination, predicted to either mimic or reverse an input gene expression signature <a href='#Citations'> [6]</a>. It integrates the library of intergrated network-based cellular signatures (LINCS) L1000 data set processed with [Characteristic Direction Method](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-79) <a href='#Citations'> [7]</a>.It currently includes information about more than a million gene expression profiles of chemically perturbed human cell lines. L1000CDS<sup>2</sup> was used to identify small molecules that would mimic or reverse the signature, using differential gene expression signatures computed using [Characteristic Direction Method](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-79).\n",
    "\n",
    "        \"\"\")\n",
    "    markdown_cell_24 = nbformat.v4.new_markdown_cell(markdown_cell_24_content)\n",
    "    nb.cells.append(markdown_cell_24)\n",
    "\n",
    "\n",
    "    source_16=textwrap.dedent(\"\"\"\\\n",
    "\n",
    "    import retrieve\n",
    "    smq_mimic=retrieve.small_molecule_query(top_500,bottom_500, mimic=True)[0]\n",
    "    smq_reverse=retrieve.small_molecule_query(top_500,bottom_500, mimic=False)[0]\n",
    "\n",
    "    df = pd.DataFrame([smq_mimic,smq_reverse])\n",
    "\n",
    "    df.index=['Small molecules predicted to mimic the signature','Small molecules predicted to reverse the signature']\n",
    "    df.columns=['Links to Small Molecule Query results']\n",
    "    def make_clickable(val):\n",
    "        return '<a href=\"{}\">{}</a>'.format(val,val)\n",
    "    df.style.format(make_clickable)\n",
    "    \"\"\")\n",
    "    code_cell_16 = nbformat.v4.new_code_cell(source=source_16)\n",
    "    nb.cells.append(code_cell_16)\n",
    "\n",
    "    markdown_cell_25_content = textwrap.dedent(\"\"\"\\\n",
    "        #### Table 4| Links to  Small Molecule Query results\n",
    "        The table includes the links to Small Molecule Query results. The resulting small molecules are reported, along with the cell line perturbed, dosage and time of exposure.\n",
    "        \"\"\")\n",
    "    markdown_cell_25 = nbformat.v4.new_markdown_cell(markdown_cell_25_content)\n",
    "    nb.cells.append(markdown_cell_25) \n",
    "\n",
    "\n",
    "\n",
    "    source_19=textwrap.dedent(\"\"\"\\\n",
    "    \n",
    "    output=pd.DataFrame()\n",
    "    output_mimic=retrieve.small_molecule_query(top_500,bottom_500, mimic=True)[1].get('topMeta')\n",
    "    output_reverse=retrieve.small_molecule_query(top_500,bottom_500, mimic=False)[1].get('topMeta')\n",
    "\n",
    "\n",
    "    pert_id_mimic = [item.get('pert_id') for item in output_mimic]  \n",
    "    pert_desc_mimic = [item.get('pert_desc') for item in output_mimic]   \n",
    "\n",
    "    pert_id_reverse = [item.get('pert_id') for item in output_reverse]  \n",
    "    pert_desc_reverse = [item.get('pert_desc') for item in output_reverse ]  \n",
    "    pert_overlap_reverse = [item.get('score') for item in output_reverse]\n",
    "    pert_overlap_mimic= [item.get('score') for item in output_mimic]\n",
    "\n",
    "    df_pert_desc_reverse = pd.DataFrame({'Small molecule description':pert_desc_reverse,'Small molecule ID': pert_id_reverse,'Overlap score':pert_overlap_reverse})\n",
    "    df_pert_desc_reverse = df_pert_desc_reverse[['Small molecule description','Small molecule ID','Overlap score']]\n",
    "\n",
    "\n",
    "    df_pert_desc_mimic = pd.DataFrame({'Small molecule description':pert_desc_mimic,'Small molecule ID': pert_id_mimic,'Overlap score':pert_overlap_mimic})\n",
    "    df_pert_desc_mimic = df_pert_desc_reverse[['Small molecule description','Small molecule ID','Overlap score']]\n",
    "    qgrid.show_grid(df_pert_desc_mimic)\n",
    "    \"\"\")\n",
    "    code_cell_19 = nbformat.v4.new_code_cell(source=source_19)\n",
    "    nb.cells.append(code_cell_19)\n",
    "\n",
    "    markdown_cell_20_content = textwrap.dedent(\"\"\"\\\n",
    "         #### Table 5 | Small Molecules predicted to mimic the input gene expression signature\n",
    "        The table shows the list of small molecules that are predicted to mimic the input gene expression signature\n",
    "        \"\"\")\n",
    "    markdown_cell_20 = nbformat.v4.new_markdown_cell(markdown_cell_20_content)\n",
    "    nb.cells.append(markdown_cell_20)\n",
    "\n",
    "\n",
    "\n",
    "    source_21=textwrap.dedent(\"\"\"\\\n",
    "    df_pert_desc_reverse.index=range(1,len(df_pert_desc_reverse)+1)\n",
    "    qgrid.show_grid(df_pert_desc_reverse)\n",
    "    \n",
    "    \"\"\")\n",
    "    code_cell_21 = nbformat.v4.new_code_cell(source=source_21)\n",
    "    nb.cells.append(code_cell_21)\n",
    "\n",
    "    markdown_cell_30_content = textwrap.dedent(\"\"\"\\\n",
    "         #### Table 6 | Small Molecules predicted to reverse the imput gene expression signature\n",
    "        The table shows the list of small molecules that are predicted to reverse the input gene expression signature\n",
    "        \"\"\")\n",
    "    markdown_cell_30 = nbformat.v4.new_markdown_cell(markdown_cell_30_content)\n",
    "    nb.cells.append(markdown_cell_30)\n",
    "    \n",
    "    \n",
    "    markdown_cell_18 = nbformat.v4.new_markdown_cell(\"<a id='Citations'></a>\")\n",
    "    nb.cells.append(markdown_cell_18)  \n",
    "\n",
    "    markdown_cell_citations=(\"### Citations\")\n",
    "    \n",
    "    markdown_cell_citations=nbformat.v4.new_markdown_cell(markdown_cell_citations)\n",
    "    nb.cells.append(markdown_cell_citations)\n",
    "    \n",
    "    markdown_cell_28_content = textwrap.dedent(\"\"\"\\\n",
    "    \n",
    "    1.  Perez, F.,  Granger, B. E. (2015). Project Jupyter: Computational narratives as the engine of collaborative data science. Technical Report. Technical report, Project Jupyter. \n",
    "    2.  Clark, N. R.,  Ma'ayan, A. (2011). Introduction to statistical methods to analyze large data sets: Principal components analysis. Science signaling, 4(190), tr3.\n",
    "    3.  Fernandez, N. F., Gundersen, G. W., Rahman, A., Grimes, M. L., Rikova, K., Hornbeck, P.,  Ma'ayan, A. (2017). Clustergrammer, a web-based heatmap visualization and analysis tool for high-dimensional biological data. Scientific Data, 4, sdata2017151.\n",
    "    4.  Smyth, G. K., Ritchie, M., Thorne, N., Wettenhall, J., Shi, W., Hu, Y. (2002). Limma: Linear Models for Microarray and RNA-Seq Data User's Guide.\n",
    "    5.  Kuleshov MV, Jones MR, Rouillard AD, et al. Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic Acids Research. 2016;44(Web Server issue):W90-W97. doi:10.1093/nar/gkw377.\n",
    "    6.  Duan, Q. L1000CDS2: LINCS L1000 Characteristic Direction Signatures Search Engine. npj Systems Biology and Applications. 2016.\n",
    "    7.  Clark, N. R., Hu, K. S., Feldmann, A. S., Kou, Y., Chen, E. Y., Duan, Q.,  Ma'ayan, A. (2014). The characteristic direction: a geometrical approach to identify differentially expressed genes. BMC bioinformatics, 15(1), 79.\"\"\")\n",
    "    \n",
    "    markdown_cell_28 = nbformat.v4.new_markdown_cell(markdown_cell_28_content)\n",
    "    nb.cells.append(markdown_cell_28)\n",
    "\n",
    "\n",
    "    # executing and saving the notebook\n",
    "    from nbconvert.preprocessors.execute import ExecutePreprocessor\n",
    "\n",
    "    pp = ExecutePreprocessor()\n",
    "    pp.timeout = 30  # seconds\n",
    "    pp.interrupt_on_timeout = True\n",
    "\n",
    "    nb_executed, resources = pp.preprocess(nb, resources={})\n",
    "\n",
    "    nb_name = \"Automated_Notebook\" + \"_\" + sample_id + \".\" + \"ipynb\"\n",
    "\n",
    "    with open(nb_name, 'w') as f:\n",
    "        f.write(nbformat.v4.writes_json(nb_executed))\n",
    "\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12+"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
